streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java

```diff
@@ -0,0 +1,859 @@
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.apache.kafka.streams.KafkaStreams.State.ERROR;
+import static org.apache.kafka.streams.KafkaStreams.State.REBALANCING;
+import static org.apache.kafka.streams.KafkaStreams.State.RUNNING;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+@Category({IntegrationTest.class})
+public class KStreamRepartitionIntegrationTest {
+    private static final int NUM_BROKERS = 1;
+    private static final AtomicInteger TEST_NUM = new AtomicInteger(0);
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(NUM_BROKERS);
+
+    private final String inputTopic = "input-topic-" + TEST_NUM.get();
+    private final String outputTopic = "output-topic-" + TEST_NUM.get();
+    private final String applicationId = "kstream-repartition-stream-test-" + TEST_NUM.get();
+
+    private Properties streamsConfiguration;
+    private List<KafkaStreams> kafkaStreamsInstances;
+
+    @Before
+    public void before() throws InterruptedException {
+        CLUSTER.createTopic(inputTopic, 4, 1);
+        CLUSTER.createTopic(outputTopic, 1, 1);
+
+        streamsConfiguration = new Properties();
+        kafkaStreamsInstances = new ArrayList<>();
+
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsConfiguration.put(StreamsConfig.TOPOLOGY_OPTIMIZATION, StreamsConfig.OPTIMIZE);
+    }
+
+    @After
+    public void whenShuttingDown() throws IOException {
+        kafkaStreamsInstances.stream()
+                             .filter(Objects::nonNull)
+                             .forEach(KafkaStreams::close);
+
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+        TEST_NUM.incrementAndGet();
```

Comment:
With parallel test runners, would it be better to call this as first line in `before()` method (and use the returned value instead of calling `get()` in addition -- otherwise, we might get multiple different numbers per test run)? Also wondering if we should assign the topic names that use the counter within before?

For a clean isolation, it might also be good to add the test number to the `application.id`

Reply:
`application.id` already has test number.

Will do as you suggested.